ironfish/src/wallet/scanner/queue.ts

```diff
@@ -0,0 +1,124 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+const EMPTY = Symbol('EMPTY')
+
+function assertValidMaxSize(maxSize: number) {
+  if (maxSize <= 0) {
+    throw new Error('maxSize cannot be greater than 0')
+  }
+}
+
+function assertNotEmpty<T>(x: T | typeof EMPTY): asserts x is T {
+  if (x === EMPTY) {
+    throw new Error('Expected item not to be empty')
+  }
+}
+
+/**
+ * A fixed-size, async queue, implemented on top of a circular buffer.
+ */
+export class Queue<Item> {
```

Comment:
We have a priority queue in `/src/utils`, wonder if this should be in the same directory

Reply:
Sounds good to me, I can move it there

Reply:
My only feedback is that I would rename this to AsyncQueue since Javascript already has a built in queue (Array).

When I saw this I didn't understand why this is re-implementing the Javascript array since the default array in javascript is a queue that works well. I expected this to be an Async wrapper on top of array. I think the circular buffer is a fine optimization, but not necessary because most of the time you aren't dealing with throughput and allocations but waiting on async operations to clear the queue.

Reply:
Ok to rename this to `AsyncQueue`.

I'd like to point out though that this is a FIFO queue; an array could only be used efficiently for a LIFO queue. I've seen in our code base that we sometimes use arrays as FIFO queues using `.shift()`. For large queues of N elements, this has the side effect that N-1 elements have to be read and re-written for every `pop()` operation. This has a huge impact on the performance of the queue and anything using it.